serdo.8
.TH serdo 1
.SH NAME
serdo \- run commands serially
.SH SYNOPSIS
.B serdo
.I [-c]
.I filename

.SH DESCRIPTION
.B serdo
will open the file given by the command line argument and serially
execute all the commands in it.  If a command fails, the whole batch job
is aborted (unless -c is given as first paramter on the serdo command
line).

serdo understands the \fBcd\fR and \fBexport\fR sh(1) built-ins (no
loops, no ~user expansion, no $FOO expansion, no backticks).

serdo is very limited by design, but it is nice to have if you just want
to run a few ifconfig, ip, route commands in sequence.  serdo will
return the exit code of the last command it ran, 0 if none were given.

.SH AUTHOR
minit was written by Felix von Leitner and can be downloaded from
.I http://www.fefe.de/minit/

.SH "SEE ALSO"
msvc(8), sh(1)


CHANGES
0.10:
  add sample script for /etc/minit/ctrlaltdel/run as
    control/ctrlaltdel-run (Florian Westphal)
  add killall5, needs /proc on Linux though (Florian Westphal)
  add shutdown (Bernd Wachter)
  add -Wall and -W and kill resulting compiler warnings
  several fixes (Bernd Wachter)
  add update support - tell minit to execve something else, a newer
    version of itself, preferably (Florian Westphal)
  also add minit-update program to download state from old minit and
    upload state to new minit after update (Florian Westphal)
  #include minit.h and use MINITROOT consistently (Lordy)
  Don't assume MINITROOT is 10 bytes long (Erich Schubert)
  add man pages!! (Erich Schubert)
  add a way to ask minit for dependencies
  msvc -H will now dump the ten least recently spawned processes.
    This is useful to see which process is looping.
  fix bug in msvc killing process group if readpid failed
    (Erich Schubert)
  You can now blacklist services temporarily by passing -servicename on
    the kernel command line.  The first time they are to be started will
    then be skipped.  Use this, for example, to not start the dhcp
    client when your notebook is not plugged in a network.
  add serdo (execute batch, no shell features whatsoever)
  make "msvc servicename" output easier to parse if not outputting to a tty
  require libowfat instead of shipping half of it ourselves
  replace malloc with alloca; minit is now 6644 bytes on i386.
  add usage for serdo (Nikola Vladov)
  fix msvc -d followed by msvc -h (would send SIGHUP to pid 1, i.e. kill
    minit; found by Paul Fox)
  added small special-purpose malloc using a static buffer before
    calling the real malloc (Nikola Vladov)

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  For stopped services, minit did not record the time when it finished,
    but when it was started.
  add hard-reboot from Tommi Virtanen.  He also asked me to include an
    explicit license statement and sent a patch to make debian packaging
    easier (deb files can't include FIFOs).
  Olaf: add write_proc.
  Olaf: lines in depends can now be commented out with a #
  Olaf: add a method to clear dead services (mark as terminated)
  minit now checks whether a PID actually exists before accepting msvc -P
    this removes a race condition when the forked service terminates
    before pidfilehack notifies minit of the PID
  msvc now accepts more than one service after -o, -d, ...
  msvc now accepts /etc/minit/sshd instead of sshd as service name
  better error handling in msvc (more and better error messages)
  minit now has a heuristic to detect time zone changes and still
    have msvc report the correct elapsed time since process start/end
    (don't look at the source code, it's ultra-kludgey)
  add msvc -g (get PID for shell scripts)

0.8:
  call waitpid repeatedly until it returns "no children".
  This reaps zombies faster.

0.7:
  found and fixed the bug that made the logging support not work when
    minit was running as PID 1.
  saw that the shutdown from sysvinit actually does what it's supposed
    to do, even with minit, if one uses the -n mode.  It should be quite
    easy to write a nice and clean shutdown for minit now.
  if a service depends on a service that already finished, don't start
    it again.

0.6.1:
  fixed msvc and minit handling of msvc -P.
  pidfilehack has now actually been tested and works with ssh.

0.6:
  add tty and session leadership magic and I actually got getty to work
    as init on my laptop, i.e. in production use!
  add msvc man page.

0.5.1:
  cut&paste error, set wrong variable to 0 for keyboard request.
    Thanks, Schulti.

0.5:
  add "log" subdirectory (see README).
  get rid of stdio and *printf and switch to routines from libowfat.

0.4:
  add pidfilehack.  It will take the first argument as filename for the
  PID file (i.e. /var/run/sshd.pid), the second argument as service name
  for minit (i.e. ssh) and the rest of the command line is run without
  path searching in a child process.  The parent process will then
  try to open the pidfile several times (for up to 30 seconds), read the
  PID off it and run msvc to tell minit the PID of the service.  minit
  can then track sshd and restart it if necessary.  This gross hack does
  not work for daemons that do not write a PID file, but there are not
  many of those, fortunately.


serdo.c
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <ctype.h>
#include <errmsg.h>
#include <sys/wait.h>
#include <str.h>
#include <byte.h>

extern char **environ;

#define MAXENV 256
char* envp[MAXENV+2];
int envc;

int continueonerror;

int envset(char* s) {
  int i,l;
  if (s[l=str_chr(s,'=')]!='=') return -1;
  ++l;
  for (i=0; i<envc; ++i)
    if (byte_equal(envp[i],l,s)) {
      envp[i]=s;
      return 0;
    }
  if (envc<MAXENV) {
    envp[envc]=s;
    envp[++envc]=0;
    return 0;
  }
  return -1;
}

int spawn(char** argv, int last) {
  int i;
  if (str_equal(argv[0],"cd")) {
    if (chdir(argv[1])==-1) {
      carpsys("chdir failed");
      return -1;
    }
    return 0;
  } else if (str_equal(argv[0],"export")) {
    for (i=1; argv[i]; ++i) envset(argv[i]);
    return 0;
  }
  if (!last) {
    if ((i=fork())==-1) diesys(1,"cannot fork");
  } else i=0;
  if (!i) {
    /* child */
    environ=envp;
    _exit(execvp(argv[0],argv));
  }
  if (waitpid(i,&i,0)==-1) diesys(1,"waitpid failed");
  if (!WIFEXITED(i))
    return -1;
  return WEXITSTATUS(i);
}

int run(char* s,int last) {
  int i,spaces;
  char** argv,**next;;
  for (i=spaces=0; s[i]; ++i) if (s[i]==' ') ++spaces;
  next=argv=alloca((spaces+1)*sizeof(char*));
  while (*s) {
    while (*s && isspace(*s)) ++s;
    if (*s=='"') {
      ++s;
      *next=s;
      while (*s && s[-1] != '\\' && *s != '"') ++s;
      if (!*s) {
	--*next;
	break;
      }
      *s=0;
      ++s;
    } else if (*s=='\'') {
      ++s;
      *next=s;
      while (*s && s[-1] != '\\' && *s != '\'') ++s;
      if (!*s) {
	--*next;
	break;
      }
      *s=0;
      ++s;
    } else {
      *next=s;
      while (*s && *s!=' ' && *s!='\t') ++s;
      if (!*s) break;
      *s=0;
      ++s;
    }
    ++next;
  }
  *++next=0;

  return spawn(argv,last);
}

int execute(char* s) {
  char* start;
  int r;
  r=0;
  while (*s) {
    int last;
    while (isspace(*s)) ++s;
    if (*s == '#') {
      while (*s && *s != '\n') ++s;
      continue;
    }
    start=s;

    while (*s && *s != '\n') ++s;
    if (*s) {
      char* tmp;
      *s=0;
      ++s;
      for (tmp=s; *tmp; ++tmp)
	if (!isspace(*tmp) && *tmp=='#') {
	  for (tmp=s+1; *tmp && *tmp!='\n'; ++tmp) ;
	} else break;
      last=(*tmp==0);
    } else
      last=1;
    r=run(start,last);
    if (r!=0 && !continueonerror)
      break;
  }
  return r;
}

int batch(char* s) {
  struct stat ss;
  int fd=open(s,O_RDONLY);
  char* map;
  if (fd==-1) diesys(1,"could not open ",s);
  if (fstat(fd,&ss)==-1) diesys(1,"could not stat ",s);
  if (ss.st_size>32768) die(1,"file ",s," is too large");
  map=alloca(ss.st_size+1);
  if (read(fd,map,ss.st_size)!=(long)ss.st_size) diesys(1,"read error");
  map[ss.st_size]=0;
  close(fd);

  return execute(map);
}

int main(int argc,char* argv[],char* env[]) {
  int r;
  (void)argc;
  if (argc<2) die(1,"usage: serdo [-c] filename");
  errmsg_iam("serdo");
  for (envc=0; envc<MAXENV && env[envc]; ++envc) envp[envc]=env[envc];
  envp[envc]=0;
  if (str_equal(argv[1],"-c")) {
    continueonerror=1;
    ++argv;
  }
  while (*++argv) {
    if ((r=batch(*argv)))
      return r;
  }
  return 0;
}
